io/epoll.cpp

```diff
@@ -117,54 +117,57 @@ class EventEngineEPoll : public MasterEventEngine, public CascadingEventEngine,
-        if (e.interests & entry.interests) {
-            if (((e.interests & entry.interests & EVENT_READ) &&
-                 (entry.reader_data != e.data)) ||
-                ((e.interests & entry.interests & EVENT_WRITE) &&
-                 (entry.writer_data != e.data)) ||
-                ((e.interests & entry.interests & EVENT_ERROR) &&
-                 (entry.error_data != e.data))) {
+        auto eint = e.interests & entry.interests;
+        if (((eint & EVENT_READ)  && (entry.reader_data != e.data)) ||
+            ((eint & EVENT_WRITE) && (entry.writer_data != e.data)) ||
+            ((eint & EVENT_ERROR) && (entry.error_data  != e.data))) {
                 LOG_ERROR_RETURN(EALREADY, -1, "conflicted interest(s)");
-            }
         }
 
         if (e.interests & EVENT_READ) entry.reader_data = e.data;
         if (e.interests & EVENT_WRITE) entry.writer_data = e.data;
         if (e.interests & EVENT_ERROR) entry.error_data = e.data;
-        auto eint = entry.interests & (EVENT_READ | EVENT_WRITE | EVENT_ERROR);
-        auto op = eint ? EPOLL_CTL_MOD : EPOLL_CTL_ADD;
-        if (op == EPOLL_CTL_MOD &&
-            (e.interests & ONE_SHOT) != (entry.interests & ONE_SHOT)) {
-            LOG_ERROR_RETURN(
-                EINVAL, -1,
-                "do not support ONE_SHOT on no-oneshot interested fd");
-        }
-        auto x = entry.interests |= e.interests;
-        x &= (EVENT_READ | EVENT_WRITE | EVENT_ERROR);
-        // since epoll oneshot shows totally different meanning of ONESHOT in
-        // photon all epoll action keeps no oneshot
+        eint = entry.interests;
+        entry.interests |= e.interests;
+        auto x = entry.interests & EVENT_RWE;
         auto events = evmap.translate_bitwisely(x);
-        return ctl(e.fd, op, events);
+        if (e.interests & ONE_SHOT) {
+            events |= EPOLLONESHOT;
+            auto op = likely(eint & ONE_SHOT) ? EPOLL_CTL_MOD : EPOLL_CTL_ADD;
+            auto ret = ctl(e.fd, op, events);
+            return likely(ret >= 0) ? ret :
+                       ctl(e.fd, EPOLL_CTL_ADD, events);
+        } else {
+            auto op = (eint & EVENT_RWE) ? EPOLL_CTL_MOD : EPOLL_CTL_ADD;
+            return ctl(e.fd, op, events);
+        }
     }
     virtual int rm_interest(Event e) override {
         if (e.fd < 0 || (size_t)e.fd >= _inflight_events.size())
             LOG_ERROR_RETURN(EINVAL, -1, "invalid file descriptor ", e.fd);
+        if (unlikely(!e.interests)) return 0;
         auto& entry = _inflight_events[e.fd];
-        auto intersection = e.interests & entry.interests &
-                            (EVENT_READ | EVENT_WRITE | EVENT_ERROR);
+        auto intersection = e.interests & entry.interests & EVENT_RWE;
         if (intersection == 0) return 0;
 
-        auto x = (entry.interests ^= intersection) &
-                 (EVENT_READ | EVENT_WRITE | EVENT_ERROR);
+        entry.interests ^= intersection;
```

Comment:
```suggestion
        entry.interests &= ~intersection;
```
Easier to understand

Reply:
added comments